Add playlist name and video name in playlist sharing content

snaik20 opened the pull request:
<!-- Hey there. Thank you so much for improving NewPipe, and filling out the details. Having roughly the same layout helps everyone considerably :)-->

#### What is it?
- [ ] Bugfix (user facing)
- [x] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
<!-- While bullet points are the norm in this section, feel free to write free-form text instead of a list -->
Add playlist name and video name in playlist sharing content

- Currently, only a list of videos separated by newline are added in
  the share content.
- This makes it difficult to identify a specific video in a list of
  Urls.
- Used string resources for the sharing content formats.
- Added a confirmation dialog for users to choose between sharing
  playlist formats.
- Added Playlist name as the header and corresponding video name for
  each video Url in following format.
```
My Playlist
- Music1: https://media-url1
- Music2: https://media-url2
- Music3: https://media-url3
```

#### Before/After Screenshots/Screen Record
<!-- If your PR changes the app's UI in any way, please include screenshots or a video showing exactly what changed, so that developers and users can pinpoint it easily. Delete this if it doesn't apply to your PR.-->
| Before | After | Confirmation Dialog |
| --- | --- | --- |
| <img src="https://github.com/TeamNewPipe/NewPipe/assets/87667048/fa7b27b2-91d6-491d-8f93-a8b447263de6" width=270 height=580> | <img src="https://github.com/TeamNewPipe/NewPipe/assets/87667048/e7ddde18-1d3a-4ccb-8c35-f5bf14e48bec" width=270 height=580> | <img src="https://github.com/TeamNewPipe/NewPipe/assets/87667048/c073c41e-8b49-461b-95b2-cab5e7ffb692" width=270 height=580> |

#### Video demo:
https://github.com/TeamNewPipe/NewPipe/assets/87667048/09962034-146a-4fa6-acbc-5bc58cba2d10

#### Fixes the following issue(s)
<!-- Prefix issues with "Fixes" so that GitHub closes them when the PR is merged (note that each "Fixes #" should be in its own item). Also add any other relevant links. -->
- Supersedes and closes #9060
- Fixes #9025 

#### APK testing
<!-- Use a new, meaningfully named branch. The name is used as a suffix for the app ID to allow installing and testing multiple versions of NewPipe, e.g. "commentfix", if your PR implements a bugfix for comments. (No names like "patch-0" and "feature-1".)  -->
<!-- Remove the following line if you directly link the APK created by the CI pipeline. Directly linking is preferred if you need to let users test.-->
The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR. You can find more info and a video demonstration [on this wiki page](https://github.com/TeamNewPipe/NewPipe/wiki/Download-APK-for-PR).

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).
**Note:** I am aware of the NewPipe rewrite, and I would be open to updating my PR as per requirements.

TobiGr commented:
Thank you for the changes. Please add screenshots of the share sheets showing the state before and after your PR.

snaik20 commented:
@TobiGr Thanks for your inputs, updated the PR with screenshots.

TobiGr commented:
Thanks!
Sometimes, people just need to URLs to be exported. Please add a preference to the ContentSettings which allows the user to export a list of URLs only.

TobiGr commented:
@Stypox just commented on Matrix that a dialog before sharing is better than a setting. He is correct. So please do not add a setting but rather an AlertDialog with a checkmark which enables the user add the video title and playlist name

TobiGr commented on app/src/main/res/values/strings.xml:
```diff
@@ -811,4 +811,10 @@
     <string name="show_channel_tabs">Channel tabs</string>
     <string name="show_channel_tabs_summary">What tabs are shown on the channel pages</string>
+    <string name="share_playlist">Share Playlist</string>
+    <string name="share_playlist_with_details_message">Share playlist with details such as playlist name and video titles Or a simply share a list of video links</string>
+    <string name="share_playlist_with_details">Share with details</string>
+    <string name="share_playlist_with_list">Share list</string>
```
```suggestion
    <string name="share_playlist_with_details_message">Share playlist with details such as playlist name and video titles or as a simple list of video URLs</string>
    <string name="share_playlist_with_details">Share with details</string>
    <string name="share_playlist_with_list">Share URL list</string>
```

TobiGr commented on app/src/main/res/values/settings_keys.xml:
```diff
@@ -24,6 +24,8 @@
     <string name="clear_queue_confirmation_key">clear_queue_confirmation_key</string>
     <string name="ignore_hardware_media_buttons_key">ignore_hardware_media_buttons_key</string>
+    <string name="share_playlist_with_details_key">share_playlist_with_details_key</string>
+    <string name="share_playlist_with_details_can_show_dialog_key">share_playlist_with_details_can_show_dialog_key</string>
```
I think this name is better to understand what the key is used for. I think the usage needs to be inverted though.
```suggestion
    <string name="remember_local_playlist_sharing_option">remember_local_playlist_sharing_option</string>
```

TobiGr commented on app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java:
```diff
@@ -374,16 +385,27 @@ public boolean onOptionsItemSelected(final MenuItem item) {
-                        .map(StreamEntity::getUrl)
+                        .map(streamEntity -> {
+                            if (shouldSharePlaylistDetails) {
+                                return String.format("- %s: %s",
```
I think we will need to allow the localization of this string for LTR countries. Please extract it into a string res.

TobiGr commented on app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java:
```diff
@@ -841,5 +863,40 @@ private PlayQueue getPlayQueue(final int index) {
+                                    !dialogBinding.rememberChoiceCheckBox.isChecked())
+                            .putBoolean(requireContext().getString(
+                                    R.string.share_playlist_with_details_key), true)
+                            .commit();
```
```suggestion
                            .apply();
```

snaik20 replied:
I had used `commit` as we need to make the changes synchronously for the `sharePlaylist` method. I have  tested with `apply` and it seems to work properly.

TobiGr commented on app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java:
```diff
@@ -841,5 +863,40 @@ private PlayQueue getPlayQueue(final int index) {
+                                    !dialogBinding.rememberChoiceCheckBox.isChecked())
+                            .putBoolean(requireContext().getString(
+                                    R.string.share_playlist_with_details_key), false)
+                            .commit();
```
```suggestion
                            .apply();
```

TobiGr commented on app/src/main/res/xml/video_audio_settings.xml:
```diff
@@ -249,5 +249,13 @@
             android:title="@string/ignore_hardware_media_buttons_title"
             app:singleLineTitle="false"
             app:iconSpaceReserved="false" />
+
+        <SwitchPreferenceCompat
+            android:defaultValue="false"
+            android:key="@string/share_playlist_with_details_key"
+            android:summary="@string/share_playlist_with_details_settings_item"
+            android:title="@string/share_playlist"
+            app:singleLineTitle="false"
+            app:iconSpaceReserved="false" />
```
I am not sure if this is the correct place for this setting. @opusforlife2 Where would you put it?

snaik20 commented:
@TobiGr Added a confirmation dialog to the playlist sharing process, allowing users to choose between sharing playlist formats. Users can also choose to save their sharing format preference to skip the confirmation dialog. In addition, an option has been added to the settings to update the sharing preference.

TobiGr reviewed (commented):
Thank you

snaik20 commented:
Thanks @TobiGr for your time and review. Updated the changes as requested. Let us wait for @opusforlife2 on inputs for settings item.

snaik20 commented:
> There's no good place to put it, unfortunately. `Content` menu is our current catch-all for miscellaneous settings. (If someone would like to tackle #9587 it would make everyone's lives easier. _wink wink_)
> 
> But why do we need a setting for this at all? Just have the checkbox state remembered. That one extra dialogue doesn't make much of a difference.

Hi @opusforlife2,

Thanks for your feedback.

I understand your concern about the extra dialog, but I believe it's important to give users the option to choose how they want to share their playlists. Some users may prefer the convenience of having a checkbox that remembers their preference, while others may want to be more deliberate about how they share their playlists.

That said, I'm open to other suggestions. Here are a few options:
 - We could remove the setting item and display the dialog every time a user shares a playlist.
 - We could remove the dialog and only keep the setting, and let the user know where to find the setting item to change the default behavior.

I'm happy to discuss these options further with you, or to implement any other solution that you think would be better.

Regarding #9587, I agree that it would be great if someone could tackle it. I would be happy to look into it.

Thanks

snaik20 commented:
https://github.com/TeamNewPipe/NewPipe/assets/87667048/5ac5ab6f-163d-4407-bd67-a078a8b55bbb



> > every time a user shares a playlist
> 
> I can't think of any obvious use cases where an average user would need to rapidly go through playlists and share them. At best this would be a one-off task. Apart from that, there could be an occasional desire to share an interesting playlist with a friend. This is why I suggested to go with:
> 
> > We could remove the setting item and display the dialog every time
> 
> Can you think of any important use cases that are frequent enough to justify an additional setting?

Hi @opusforlife2,

I think you make a good point. I can't think of any use cases where the average user would need to rapidly share playlists. I have updated the flow to always show the dialog and updated the video demo.

WDYT

snaik20 commented:
@opusforlife2 Thanks, updated.

TobiGr reviewed (approved):
Thanks!

TobiGr commented:
I fixed the wrongly formatted string resources in 0758cd6980d2cda88803930d34227b11f10d05d2

snaik20 commented:
> I fixed the wrongly formatted string resources in 0758cd6980d2cda88803930d34227b11f10d05d2

Thanks @TobiGr
